content/concepts/did-ddo.md

```diff
@@ -290,6 +301,10 @@ Example:
     {
       "type": "ipfs",
       "hash": "XXXX"
+    },
+    {
+      "type": "arweave",
+      "transactionId": "cZ6j5PmPVXCq5Az6YGcGqzffYjx2JnsnlSajaHNr20w"
```

Comment:
this looks like we are running with this bad API design, there different file objects will have different keys under them. Best practice for multiple objects with different values sharing the same schema is to generalize the keys. Like, with this change added a file object now can have either of those keys: `url`, `hash`, and now with this change `transactionId`.

So a typical TypeScript typing instead of having:

```ts
interface FileItem {
  type: string
  value: string
}
```

We then need to do this, super cumbersome and against all best practices:

```ts
interface FileItem {
  type: string
  url?: string
  hash?: string
  transactionId?: string
}
```

Then in code need to map the type to the respective key and such. And as soon as the next person adds another file integration we add more keys? 

Now is the best time to change those bad practices we already put in for URL & IPFS (although IPFS is not used in market yet) and not continue repeating them. Not really job of this PR but without a generic approach defined first, we will just add to the chaos otherwise.

Reply:
I think you raise a good point and I would be fine with this generalization. I'd have to redo a bunch of my tests 😅 but that's fine.

Reply:
@kremalicious I updated the docs per your suggestions.  Can you take a look and resolve this comment if the changes look good?

Reply:
wait, we cannot do this !
It will break all existing assets.

Going back to specs, each type of storage can have different parameters, so we could have the following typings:
```
interface UrlFile{
    type: 'url',
    url: string,
    method: string
}

interface ArwaveFile{
    type: 'arwave',
    transactionId: string
}

interface FileCoinFile{
    type: 'filecoin',
    cid: string
}

interface GraphqlFile{
    type: 'graphql',
    server: string,
    query: string
}


export interface Files {
  nftAddress: string
  datatokenAddress: string
  files: UrlFile[] | ArwaveFile[] | FileCoinFile[] | GraphqlFile
}
```

Reply:
I don't even want to mention odbc storage (databases), where you have database names, tables, clauses, etc :)

Reply:
😞 Oof. I already spent the last day and a half updating provider and ocean.py to use the generalized `value` key. But I see @alexcos20's point. I'll revert my latest changes to make the interfaces specialized again.

Reply:
Unfortunately we can't have a generic interface for all, they are too different. We can have one interface with a lot of optional props, that is not a good practice. It's way better and pretty common in oop what Alex suggested. With a minor correction:
```
export interface Files {
  nftAddress: string
  datatokenAddress: string
  files: (UrlFile ArweaveFile | FileCoinFile| GraphqlFile | ChainLinkFile)[]
}
```

And create a base interface for all
```
interface BaseFile {
 type : 'url' | 'chainlink' | ...
}

Reply:
@mihaisc Agreed, I will do like you've said.  ocean.py also uses interface inheritance like this.

Reply:
@kremalicious ?

Reply:
all good points, sounds good! Wasn't aware of value monsters like `ChainLinkFile` and while we could still generalize one "main value" for each file type, as soon as more values are required, code is actually more readable when keys have specific names